bitcoind-regtest/src/Bitcoin/Core/Regtest/Generator.hs

```diff
@@ -85,41 +86,109 @@ outputsPerSeries =
+    , getMeanFeeRate :: BlockHeight -> Word64
+    -- ^ Mean fee rate at height
+    }
+
+data GeneratorStatus
+    = Starting
+    | Running
+    | Paused
+    | Stopped
+    deriving (Eq, Show)
+
+data GeneratorState = GeneratorState
+    { status :: GeneratorStatus
+    , balance :: Word64
+    }
+    deriving (Eq, Show)
+
+data GeneratorHandle = GeneratorHandle
+    { getGeneratorState :: IO GeneratorState
+    , pauseGenerator :: IO ()
+    , resumeGenerator :: IO ()
+    , stopGenerator :: IO ()
+    , makePayment :: [(Text, Word64)] -> IO ()
+    , reconfigure :: GeneratorConfig -> IO ()
+    , generatorAsync :: Async ()
+    -- ^ Use to wait on the generator process
+    }
 
 -- | Generate many transactions per block with a certain mean fee rate
 generateWithTransactions ::
     Manager ->
     NodeHandle ->
-    -- | Block interval (seconds per block)
-    Int ->
-    -- | Optional funding behavior
-    Maybe Funding ->
-    -- | Mean fee rate at height
-    (BlockHeight -> Word64) ->
-    IO ()
-generateWithTransactions mgr nodeHandle blockInterval funding getMeanFeeRate = do
-    initFloodState <- run initialize
-
-    (async >=> link) . run . RPC.withWallet miningWallet $ case funding of
-        Just (Faucet getExternalAddress) ->
-            waitRepeat 1 $ do
-                balance <- getBalance
-                when (balance >= faucetPayment) $
-                    liftIO getExternalAddress >>= mapM_ (fundAddress faucetPayment)
-        Just (OneTime payouts) -> mapM_ (uncurry handlePayout) payouts
-        _ -> pure ()
-
-    run . flip evalStateT initFloodState . waitRepeat blockInterval $ do
-        txFlood getMeanFeeRate
-        getSeriesCount
-            >>= fmap snd . lift . generatorGenerate
-            >>= mapM_ pushRoots
+    GeneratorConfig ->
+    IO GeneratorHandle
+generateWithTransactions mgr nodeHandle initGeneratorConfig = do
+    refState <- newIORef GeneratorState{status = Starting, balance = 0}
+    refConfig <- newIORef initGeneratorConfig
+    refPaymentQueue <- newIORef mempty
+    refStatusRequest <- newIORef Nothing
+
+    let getStatus = status <$> readIORef refState
+        setStatus status =
+            atomicModifyIORef' refState $
+                \genState -> (genState{status}, status)
+        requestStatus =
+            atomicModifyIORef' refStatusRequest . updateRef . const . Just
+        onStatusRequest theStatus = do
+            -- Clear the existing status request
+            atomicModifyIORef' refStatusRequest . updateRef $ const Nothing
+            setStatus theStatus
+        mainLoop = do
+            liftIO
+                (readIORef refStatusRequest)
+                >>= liftIO . maybe getStatus onStatusRequest
+                >>= \case
+                    -- We are done
+                    Stopped -> pure ()
+                    -- Poll rather than doing a round
+                    Paused -> liftIO (threadDelay oneSec) >> mainLoop
+                    _ -> do
```

Comment:
not sure I follow `GeneratorStatus` transitions, but doesn't this run twice? for both when `Starting` and `Running` the generator? is that a problem?

Reply:
The `mainLoop` function defines a single round of computation for the generator, depending on the current state.  We run it over and over until we reach the `Stopped` state, where we exit.


There is no way to transition into `Starting` from another state.  So in practice, this logic only covers the `Stopped`, `Paused`, and `Running` cases.